nukkit/src/main/java/network/ycc/waterdog/nukkit/WaterdogAPI.java

```diff
@@ -42,23 +45,51 @@ public void onEnable() {
 		this.conf = new WaterdogToolsConfig(this, "config.yml");
 		if (this.conf.isProxyFirewallEnabled()) {
 			this.getLogger().info("[Firewall] Enabling Network-Firewall...");
+			List<String> allowedIPs = this.conf.getProxyAddresses();
+			Map<UUID, Player> online = this.getServer().getOnlinePlayers();
+			if (online != null && !online.isEmpty()) {
+				for (Player p : online.values()) {
+					try {
+						if (!allowedIPs.contains(p.getAddress())) {
+							this.getLogger().info("[Firewall] Kicking non-proxy connected Player " + p.getName());
+							p.kick("Please connect over the proxy.");
+						}
+					} catch (Exception e) {
+						continue;
+					}
+				}
+			}
 			if (!this.conf.isPrimaryFirewallDisabled()) {
+				boolean wasAlreadyregistered = false;
 				for (SourceInterface sourceInterface : this.getServer().getNetwork().getInterfaces()) {
 					if (sourceInterface instanceof RakNetInterface) {
-						sourceInterface.shutdown();
-						this.getServer().getNetwork().unregisterInterface(sourceInterface);
+						if (sourceInterface.getClass().getName().equals(ProtectedRakNetInterface.class.getName())) {
```

Comment:
Can't use instanceof. This should still work in newer java versions hopefully-

Reply:
Why not?

Reply:
Because the class this loads is in plugin space and therefore not the same after a ClassLoader reload

Reply:
I see